### PR TITLE
[mini-browser] Better handle iframe errors

### PR DESCRIPTION
**Pull request by jankeromnes**
Today, we don't handle iframe errors very well in Theia, as most of the time we hide them behind a forever-spinning load indicator overlay. (See #3589, #4102, https://github.com/gitpod-io/gitpod/issues/260.)

ToDo:

- [x] Hide the load indicator if the iframe sends a `'load'` event (was already the case)
- [x] Hide the load indicator if the iframe sends an `'error'` event (new with this PR)
- [x] Hide the load indicator if the iframe hasn't loaded after some time

This will make sure that we stop hiding an iframe that had errors. Most of the time, the browser will display more information on the error inside the iframe (e.g. CSP, content encoding, cross-origin, invalid address, etc.). However, in some cases (e.g. HTTP Mixed Content) the errored iframe will be completely blank.

Potential further improvements:

- [x] Insert a banner with an error message, between the Preview URL bar and the iframe, or at the bottom of the iframe
  - Showing something like "Error: iframe timed out", or the actual error if we have it
- [x] Add a fade-out animation when removing the load indicator
- ~~See if we can detect more iframe errors by polling `try { frame.contentWindow.document.readyState } catch (e) {  }`~~ (it detects errors, but also for sites that work)

### CODE REVIEW

**Comment by jankeromnes** [comment]
Ok, now I'd like to introduce an error banner, so that we can tell the user when there was a potential iframe error, but still show the iframe (because it might contain more info).

I drew some inspiration from the bottom status bar, and from the error notifications. Here are a few concepts:

screenshot 2019-02-27 at 14 55 11
screenshot 2019-02-27 at 14 55 48
screenshot 2019-02-27 at 15 05 45

Any thoughts or preferences?

I think I personally prefer the top orange bar (or maybe red for actual errors). I hope it plays well with the IDE's design language.

**Comment by svenefftinge** [comment]
I like the orange top bar most. The one that seem closer to existing design looks like a notification but is not.

**Comment by jankeromnes** [comment]
I implemented the orange top bar. Here is what it looks like in Firefox with various errors:

screenshot 2019-02-27 at 17 45 06
screenshot 2019-02-27 at 17 45 19
screenshot 2019-02-27 at 17 45 37
screenshot 2019-02-27 at 17 46 03
Note: The https://www.google.com behavior is special (blank page, `load` event, no error), but I think that's ok.

**Comment by jankeromnes** [comment]
Regarding the "timed out" message, maybe we should rephrase it to something like `Error: Loading this page is taking too long`.

E.g. if you do `F1` > "URL" > "https://baidu.cn", you'll probably see the time-out message, then Baidu loading, and then the time-out message disappears (because of the `load` event).

**Comment by jankeromnes** [comment]
> See if we can detect more iframe errors by polling `try { frame.contentWindow.document.readyState } catch (e) {  }`

This gives `SecurityError: Permission denied to access property "document" on cross-origin object` for https://github.com, https://www.google.com, https://www.asdfasdf.asdf, but also for sites that _do_ work, e.g. https://status.gitpod.io.

On the other hand, `http://perdu.com` gives `readyState = 'complete'`.

Maybe we don't need to poll this.

**Comment by jankeromnes** [comment]
Ok, this is now ready for review. @svenefftinge @kittaakos or @AlexTugarev please take a look when you have time.

Note: I tried to keep clean & small incremental commits, so it might be easier to review each commit separately.

**Comment by jankeromnes** [comment]
Or maybe @akosyakov if you have some time? 🙂

File: packages/mini-browser/src/browser/mini-browser-content.ts [hunk 1]
```diff
@@ -395,33 +400,95 @@ export class MiniBrowserContent extends BaseWidget {
...
+        this.errorBar.message.textContent = message;
+        this.errorBar.style.display = 'block';
+    }
+
+    protected hideErrorBar() {
+        this.errorBar.message.textContent = '';
+        this.errorBar.style.display = 'none';
+    }
+
+    protected maybeResetFrameBackground() {
```
Comment (kittaakos) [inline packages/mini-browser/src/browser/mini-browser-content.ts]: `ensureResetBackground`, or `maybeResetBackground` but please drop `Frame`. Thoughts?

File: packages/mini-browser/src/browser/mini-browser-content.ts [hunk 2]
```diff
@@ -395,33 +400,95 @@ export class MiniBrowserContent extends BaseWidget {
...
+
+    protected onFrameTimeout(): void {
+        clearTimeout(this.frameLoadTimeout);
+        this.maybeResetFrameBackground();
+        this.hideLoadIndicator();
+        this.showErrorBar('Error: Loading this page is taking longer than usual');
     }
 
     protected showLoadIndicator(): void {
+        this.loadIndicator.classList.remove('fade-out');
```
Comment (kittaakos) [inline packages/mini-browser/src/browser/mini-browser-content.ts]: For the sake of consistency, can you please move this next to the other CSS constants? So if someone reads the code can guess the class is defined in the same CSS file. Thanks!

File: packages/mini-browser/src/browser/mini-browser-content.ts [hunk 3]
```diff
@@ -395,33 +400,95 @@ export class MiniBrowserContent extends BaseWidget {
...
         const sandbox = (this.props.sandbox || MiniBrowserProps.SandboxOptions.DEFAULT).map(name => MiniBrowserProps.SandboxOptions[name]);
         frame.sandbox.add(...sandbox);
         this.toDispose.push(addEventListener(frame, 'load', this.onFrameLoad.bind(this)));
+        this.toDispose.push(addEventListener(frame, 'error', this.onFrameError.bind(this)));
         return frame;
     }
 
+    protected createErrorBar(): HTMLElement & Readonly<{ message: HTMLElement }> {
+        const errorBar = document.createElement('div');
+        errorBar.classList.add(MiniBrowserContent.Styles.ERROR_BAR);
```
Comment (kittaakos) [inline packages/mini-browser/src/browser/mini-browser-content.ts]: Maybe `.noselect`.

screen shot 2019-02-28 at 21 33 08

File: packages/mini-browser/src/browser/mini-browser-content.ts [hunk 4]
```diff
@@ -590,6 +657,8 @@ export class MiniBrowserContent extends BaseWidget {
                 if (this.getToolbarProps() === 'read-only') {
                     this.input.title = `Open ${url} In A New Window`;
                 }
+                clearTimeout(this.frameLoadTimeout);
+                this.frameLoadTimeout = setTimeout(this.onFrameTimeout.bind(this), 3000) as unknown as number;
```
Comment (kittaakos) [inline packages/mini-browser/src/browser/mini-browser-content.ts]: `this.frameLoadTimeout = window.setTimeout(this.onFrameTimeout.bind(this), 3000);`?

**Comment by jankeromnes** [comment]
@kittaakos Thansk a lot for the great review! 💯 

> FYI, I got only the error bar and neither the `Blocked by Content` nor the `Hmm. We're having trouble` pages.

Yes, these iframe errors are browser-specific. My screenshots are from Firefox Nightly. I believe that Chrome shows a "unhappy file" icon + a different error message when you hover on the iframe.

I guess the important thing is that users understand that something went wrong with their iframe, and get clues on how to fix it. Generally, the browser will explain the error in the iframe; our top bar makes sure users notice errors even when the iframe stays blank; and then the browser console usually has more details on what went wrong.

File: packages/mini-browser/src/browser/mini-browser-content.ts (hunk 3 above)
Comment (jankeromnes) [reply packages/mini-browser/src/browser/mini-browser-content.ts]: Thanks for this suggestion, but as a user, I hate it when apps prevent me from copying error messages (e.g. what if I want to google that error message, or file a bug?)

Do you have a strong preference for using `.noselect`?

File: packages/mini-browser/src/browser/mini-browser-content.ts (hunk 4 above)
Comment (jankeromnes) [reply packages/mini-browser/src/browser/mini-browser-content.ts]: Thanks, I will try this (I was quite confused by why TypeScript made me do all these type conversions).

File: packages/mini-browser/src/browser/mini-browser-content.ts (hunk 3 above)
Comment (kittaakos) [reply packages/mini-browser/src/browser/mini-browser-content.ts]: I do not have a preference unless it is the same everywhere.

Error bar in mini-browser:
screencast 2019-03-01 13-28-25


Error bar in Git view:
screencast 2019-03-01 13-28-51

**Comment by kittaakos** [comment]
> I believe that Chrome shows a "unhappy file" icon + a different error message when you hover on the iframe.

It shows nothing. I have tried it in Chrome.

> Generally, the browser will explain the error in the iframe

It did not explain a thing to me just that it is taking longer than usual.

> and then the browser console usually has more details on what went wrong.

I think it was the same without your changes.

**Comment by kittaakos** [comment]
+1 for squashing the commits, by the way.

File: packages/mini-browser/src/browser/mini-browser-content.ts (hunk 2 above)
Comment (jankeromnes) [reply packages/mini-browser/src/browser/mini-browser-content.ts]: Note, this is inconsistent with the similar `.theia-hidden` class (it's also [not a constant](https://github.com/jankeromnes/theia/blob/2611d94fb04dee4c0f30fca73a26822b59c36ede/packages/core/src/browser/frontend-application.ts#L196)), but I agree to make it a constant in order to make all CSS classes in `mini-browser-content.ts` consistent.

File: packages/mini-browser/src/browser/mini-browser-content.ts (hunk 3 above)
Comment (jankeromnes) [reply packages/mini-browser/src/browser/mini-browser-content.ts]: Oh wow, there is a similar error bar in the Git view too. Thanks, I didn't know about it!

Also it seems selectable, so I'll leave the Preview error bar selectable as well.

Comment (kittaakos) [reply packages/mini-browser/src/browser/mini-browser-content.ts]: > return this.doSet(context, this.memento.before);

That's the point. See the screencast. I could not select the error message from the iframe as I could from the Git view.

Comment (jankeromnes) [reply packages/mini-browser/src/browser/mini-browser-content.ts]: Additional thought, to preserve consistency, maybe we should replace:
- `(x) Error: Loading this page is taking longer than usual` with
- `(!) Loading this page is taking longer than usual`?

Comment (jankeromnes) [reply packages/mini-browser/src/browser/mini-browser-content.ts]: Then again maybe the more JS-error-like format works better for iframe errors.

**Comment by jankeromnes** [comment]
>>  I believe that Chrome shows a "unhappy file" icon + a different error message when you hover on the iframe.
> 
> It shows nothing. I have tried it in Chrome.
> 
> > Generally, the browser will explain the error in the iframe
> 
> It did not explain a thing to me just that it is taking longer than usual.

Yes, `http://` URLs will not update the iframe (so it remains transparent, but we now reset the background to white).

I also indeed forgot that for `https://github.com`, Chrome will [fire a 'load' event but not visibly explain the error](https://github.com/gitpod-io/gitpod/issues/260#issuecomment-465637310) for some reason.

But Chrome does visibly explain `https://www.google.com` and invalid URLs. Here are a the same screenshots as earlier but in Chrome:

screenshot 2019-03-01 at 14 14 43

screenshot 2019-03-01 at 14 23 04

screenshot 2019-03-01 at 14 17 10

screenshot 2019-03-01 at 14 17 55

> > and then the browser console usually has more details on what went wrong.
> 
> I think it was the same without your changes.

Yes totally. My change is just there to alert the user to possible errors, so that they think of maybe checking the browser console.

Before my change, the loading screen would stay forever, hiding all errors that are visually explained, and making you think it was Theia loading forever rather than a potential iframe error.

**Comment by jankeromnes** [comment]
I believe I have addressed every comment, except:

- https://github.com/theia-ide/theia/pull/4399/files#r261594459 (I don't understand the difference in "selectionability" of error messages)
- https://github.com/theia-ide/theia/pull/4399/files#r261594468 (I'll try to make the error message more consistent with the Git view's, just to see, but maybe I'll stick with the JS-like error format)

**Comment by jankeromnes** [comment]
I found why the error message selection is so erratic: the transparent overlay shows up on mouse drag events, thus interfering with text selection. I'll try to fix this.

Also, here are screenshots with `(!) Message` instead of `(x) Error: Message`:

screenshot 2019-03-01 at 16 25 10

screenshot 2019-03-01 at 16 28 14

Is this better?

**Comment by jankeromnes** [comment]
Alright, I believe I've addressed everything, so this PR is ready to be merged!

Many thanks again for the helpful reviews! 👍
